networktable.c

```diff
@@ -196,6 +208,93 @@ void parse_arp_cache(void)
 	dbclose();
 }
 
+// Loop over all entries in network table and unify entries by their hwaddr
+// If we find duplicates, we keep the most recent entry, while
+// - we replace the first-seen date by the earliest across all rows
+// - we sum up the number of queries of all clients with the same hwaddr
+bool unify_hwaddr(sqlite3 *db)
+{
+	// We request sets of (id,hwaddr). They are GROUPed BY hwaddr to make
+	// the set unique in hwaddr.
+	// The grouping is constrained by the HAVING clause which is
+	// evaluated once across all rows of a group to ensure the returned
+	// set represents the most recent entry for a given hwaddr
+	char* querystr = NULL;
+	int ret = asprintf(&querystr, "SELECT id,hwaddr FROM network GROUP BY hwaddr HAVING MAX(lastQuery)");
+	if(querystr == NULL || ret < 0)
+	{
+		logg("Memory allocation failed in unify_hwaddr (%i)", ret);
+		dbclose();
+		return false;
+	}
+
+	// Perform SQL query
+	sqlite3_stmt* stmt;
+	ret = sqlite3_prepare_v2(db, querystr, -1, &stmt, NULL);
+	if( ret ){
+		logg("unify_hwaddr(%s) - SQL error prepare (%i): %s", querystr, ret, sqlite3_errmsg(db));
+		dbclose();
+		return false;
+	}
+
+	// Loop until no further (id,hwaddr) sets are available
+	while((ret = sqlite3_step(stmt)) != SQLITE_DONE)
+	{
+		// Check if we ran into an error
+		if(ret != SQLITE_ROW)
+		{
+			logg("unify_hwaddr(%s) - SQL error step (%i): %s", querystr, ret, sqlite3_errmsg(db));
+			dbclose();
+			return false;
+		}
+
+		// Obtain id and hwaddr of the most recent entry for this particular client
+		const int id = sqlite3_column_int(stmt, 0);
+		const char *hwaddr = (const char *)sqlite3_column_text(stmt, 1);
+
+		// Update firstSeen with lowest value across all rows with the same hwaddr
+		dbquery("UPDATE network "\
+		        "SET firstSeen = (SELECT MIN(firstSeen) FROM network WHERE hwaddr = \'%s\') "\
+		        "WHERE id = %i;",\
+		        hwaddr, id);
+
+		// Update numQueries with sum of all rows with the same hwaddr
+		dbquery("UPDATE network "\
+		        "SET numQueries = (SELECT SUM(numQueries) FROM network WHERE hwaddr = \'%s\') "\
+		        "WHERE id = %i;",\
+		        hwaddr, id);
+
+		// Remove all other lines with the same hwaddr but a different id
+		dbquery("DELETE FROM network "\
+		        "WHERE hwaddr = \'%s\' "\
+		        "AND id != %i;",\
+		        hwaddr, id);
+	}
+
+	// Finalize statement and free query string
+	sqlite3_finalize(stmt);
+	free(querystr);
+
+	// Ensure hwaddr is a unique field
+	// Unfortunately, SQLite's ALTER TABLE does not support adding
+	// constraints to existing tables. However, we can add a unique
+	// index for the table to achieve the same effect.
```

Comment:
What do you think of changing the table to use the MAC address as the primary key?

Reply:
The `network` table has already been created for most. `ALTER TABLE` cannot be used to change the `PRIMARY KEY` and I don't think it would be worth creating a new and importing the old database table.

The current `id` column comes at no costs, as a column type `INTEGER PRIMARY KEY` is a mere alias for the `ROWID` which is always present (we're not using `AUTOINCREMENT` here).